File: ass1/q2/q2.py
import numpy as np
import pandas as pd
from scipy.stats import bartlett, levene, f_oneway, ttest_ind
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.stats.multicomp import pairwise_tukeyhsd, MultiComparison
from statsmodels.stats.multicomp import MultiComparison
from statsmodels.stats.weightstats import DescrStatsW

# Input data
data = {
    'technique': ['T1']*6 + ['T2']*7 + ['T3']*6 + ['T4']*4,
    'scores': [
         65, 87, 73, 79, 81, 69,        # T1
         75, 69, 83, 81, 72, 79, 90,    # T2
         59, 78, 67, 62, 83, 76,        # T3
         94, 89, 80, 88                 # T4
    ]
}
df = pd.DataFrame(data)

# Bartlett's test
print("\n=== Bartlett's Test ===")
techniques = [df[df.technique == br]['scores'] for br in df.technique.unique()]
bart_result = bartlett(*techniques)
print(f"Statistic: {bart_result.statistic:.3f}, p-value: {bart_result.pvalue:.3f}")

# Levene's test
print("\n=== Levene's Test ===")
lev_result = levene(*techniques)
print(f"Statistic: {lev_result.statistic:.3f}, p-value: {lev_result.pvalue:.3f}")

# Overall mean & treatment effects
print("\n=== Overall Mean ===")
overall_mean = df['scores'].mean()
print(f"Overall mean: {overall_mean:.2f}")

print("\n=== technique Means & Treatment Effects ===")
technique_means = df.groupby('technique')['scores'].mean().reset_index()
technique_means['treatment_effect'] = technique_means['scores'] - overall_mean
print(technique_means)

# One-way ANOVA
print("\n=== ANOVA Results ===")
anova = f_oneway(*techniques)
print(f"F-statistic: {anova.statistic:.3f}, p-value: {anova.pvalue:.3f}")
# Also, show statsmodels enhanced table:
model = smf.ols('scores ~ technique', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)
print(anova_table)

# Multiple Comparisons: Tukey's HSD
print("\n=== Tukey's HSD ===")
multicomp = MultiComparison(df['scores'], df['technique'])
tukey_result = multicomp.tukeyhsd()
print(tukey_result.summary())

# Fisher's LSD (pairwise t-tests, unadjusted)
print("\n=== Fisher's LSD ===")
from itertools import combinations
for b1, b2 in combinations(df.technique.unique(), 2):
    stat, p = ttest_ind(df[df.technique == b1]['scores'], df[df.technique == b2]['scores'])
    print(f"{b1} vs {b2} p-value: {p:.3f}")

# Simultaneous Confidence Intervals (Tukey)
print("\n=== Tukey Confidence Intervals ===")
for res in tukey_result._results_table.data[1:]:
    print(f"Diff: {res[2]:.2f} CI: ({res[4]:.2f}, {res[5]:.2f})")

File: ass1/q2/q2.r
# Tegan Jain
# 23CS30065
# Assignment 1
# Question 2

# Install packages
if (!require(car)) install.packages("car", dependencies = TRUE)
library(car)
if (!require(DescTools)) install.packages("DescTools", dependencies = TRUE)
library(DescTools)
if (!require(agricolae)) install.packages("agricolae", dependencies = TRUE)
library(agricolae)

# Input the data
technique <- c(
  rep("T1", 6),
  rep("T2", 7),
  rep("T3", 6),
  rep("T4", 4)
)
scores <- c(
  65, 87, 73, 79, 81, 69, # T1
  75, 69, 83, 81, 72, 79, 90, # T2
  59, 78, 67, 62, 83, 76, # T3
  94, 89, 80, 88 # T4
)
battery_data <- data.frame(technique, scores)

# Bartlett's Test
cat("\n=== Bartlett's Test (equal variance assumption) ===\n")
print(bartlett.test(scores ~ technique, data = battery_data))

# Levene's Test
cat("\n=== Levene's Test ===\n")
print(leveneTest(scores ~ technique, data = battery_data))

# Overall Mean & Treatment Effects
overall_mean <- mean(battery_data$scores)
technique_means <- aggregate(scores ~ technique, data = battery_data, FUN = mean)
technique_means$effect <- technique_means$scores - overall_mean
cat("\n=== Overall Mean ===\n")
print(overall_mean)
cat("\n=== technique Means & Treatment Effects ===\n")
print(technique_means)

# One-way ANOVA
battery_aov <- aov(scores ~ technique, data = battery_data)
cat("\n=== ANOVA Results ===\n")
print(summary(battery_aov))

# Tukey's HSD Test
cat("\n=== Tukey's HSD ===\n")
print(TukeyHSD(battery_aov))

# Scheffe's Test
cat("\n=== Scheffe's Test ===\n")
print(ScheffeTest(battery_aov))

# Fisher's LSD
cat("\n=== Fisher's LSD Test ===\n")
print(pairwise.t.test(
  battery_data$scores,
  battery_data$technique,
  p.adjust.method = "none"
))

# Duncan's Multiple Range Test
cat("\n=== Duncan's Multiple Range Test ===\n")
print(duncan.test(battery_aov, "technique"))

# Simultaneous Confidence Intervals (Tukey)
cat("\n=== Tukey Simultaneous Confidence Intervals ===\n")
print(TukeyHSD(battery_aov, conf.level = 0.95))
